ouroboros-consensus/src/consensus-testlib/Test/Ouroboros/Consensus/ChainGenerator/Slot.hs

```diff
@@ -0,0 +1,144 @@
+    basicUnsafeNew l                     = MV_S <$> MVG.basicUnsafeNew l
+    basicInitialize (MV_S mv)            = MVG.basicInitialize mv
+    basicUnsafeReplicate i x             = MV_S <$> MVG.basicUnsafeReplicate i (coerce x)
+    basicUnsafeRead (MV_S mv) i          = coerce <$> MVG.basicUnsafeRead mv i
+    basicUnsafeWrite (MV_S mv) i x       = MVG.basicUnsafeWrite mv i (coerce x)
+    basicClear (MV_S mv)                 = MVG.basicClear mv
+    basicSet (MV_S mv) x                 = MVG.basicSet mv (coerce x)
+    basicUnsafeCopy (MV_S mv) (MV_S mv') = MVG.basicUnsafeCopy mv mv'
+    basicUnsafeMove (MV_S mv) (MV_S mv') = MVG.basicUnsafeMove mv mv'
+    basicUnsafeGrow (MV_S mv) n          = MV_S <$> MVG.basicUnsafeGrow mv n
+instance VG.Vector V.Vector S where
+    basicUnsafeFreeze (MV_S mv)       = V_S <$> VG.basicUnsafeFreeze mv
+    basicUnsafeThaw (V_S v)           = MV_S <$> VG.basicUnsafeThaw v
+    basicLength (V_S v)               = VG.basicLength v
+    basicUnsafeSlice i l (V_S v)      = V_S (VG.basicUnsafeSlice i l v)
+    basicUnsafeIndexM (V_S v) i       = coerce <$> VG.basicUnsafeIndexM v i
+    basicUnsafeCopy (MV_S mv) (V_S v) = VG.basicUnsafeCopy mv v
+    elemseq (V_S v) x y               = VG.elemseq v (coerce x) y
+instance V.Unbox S
+
+-----
+
+genAsc :: R.RandomGen g => Asc -> g -> (S, g)
+genAsc asc g =
+    bool `seq` (S bool, g')
+  where
+    (q, g') = R.random g   -- note 0 <= q <= 1
+
+    bool = q < ascVal asc
+
+showS :: S -> ShowS
+showS (S bool) = showChar $ if bool then '1' else '0'
+
+-----
+
+-- | The different kinds of element counted in this library
+data E = ActiveSlotE | EmptySlotE | SlotE
+
+inverted :: Proxy Inverted
+inverted = Proxy
+
+notInverted :: Proxy NotInverted
+notInverted = Proxy
+
+-- | Many functions in this library are parameterized over a type of kind @Pol@
+--
+-- If the type is 'Inverted', then the function will treat all 'S' values as if they were first complemented.
+--
+-- The 'PreImage' family does the corresponding parameterization of 'ActiveSlotE' and 'EmptySlotE' at the type level.
+--
+-- NOTE: No 'S' value is ever actually complemented because of 'Inverted', but
+-- the functions parameterized by 'POL' will treat them as if they were.
+data Pol = Inverted | NotInverted
+
+-- | Data for 'Pol'
+class POL (pol :: Pol) where
+    -- | Make on active slot
+    mkActive :: proxy pol -> S
+    -- | Test whether @pol@ maps the given bit to one
+    test :: proxy pol -> S -> Bool
+
+-- | Every slot is either active or empty
+complementActive ::
+     proxy pol
+  -> C.Size base SlotE
+  -> C.Count base (PreImage pol ActiveSlotE) which
+  -> C.Count base (PreImage pol EmptySlotE ) which
+complementActive _pol (C.Count n) (C.Count i) = C.Count (n - i)
+
+-- | Every slot is either active or empty
+complementEmpty ::
+     proxy pol
+  -> C.Size base SlotE
+  -> C.Count base (PreImage pol EmptySlotE ) which
+  -> C.Count base (PreImage pol ActiveSlotE) which
```

Comment:
When `s ~ SlotE`, considering the complement is maybe a bit too weird ("please count how many of these slots are no slots"), so one needs some evidence that `s` is not `SlotE`. One could write something with almost the type signature you suggested (extra `proxy` argument to avoid ambiguous types):
```haskell
class Complementable s where
  complement ::
       proxy' s
    -> proxy pol
    -> C.Size base SlotE
    -> C.Count base (PreImage pol s)                     which
    -> C.Count base (PreImage pol (PreImage Inverted s)) which

instance Complementable ActiveSlotE where
  complement _ = complementActive

instance Complementable EmptySlotE where
  complement _ = complementEmpty
```
I guess the reason why it doesn't exist is that the two separate functions are easy enough to use?

Reply:
Oh, indeed, `s` would be ambiguous without the @proxy' s@ parameter. Thanks for the elaboration.

Reply:
Added a note in 049bcc4e